pybird_dev/greenfunction.py

```diff
@@ -20,26 +27,122 @@ def C(self, a):
+        #else: return (Omega0_m/a/a/a + (1.-Omega0_m))**0.5 #LCDM
+        return (self.Omega0_m/a/a/a + (1.-self.Omega0_m))**0.5 #expansion fixed to LCDM
+
+    def dHda_NC(self, a):
+        """Derivative of non-Conformal Hubble w.r.t. a, H0=1"""
+        #if self.nDGP: return (-3.*a**(-1. - 3.*(1 + w))*(1 + w)*(1.-Omega0_m) - 3*Omega0_m*a**(-4))/(2.*np.sqrt((1.-Omega0_m)*a**(-3*(1 + w)) + Omega0_m*a**(-3) + Omega_rc)) #nDGP+DE
+        #else: return 0.5*(-3.*Omega0_m/a/a/a/a)*(Omega0_m/a/a/a + (1.-Omega0_m))**(-0.5) #LCDM
+        return 0.5*(-3.*self.Omega0_m/a/a/a/a)*(self.Omega0_m/a/a/a + (1.-self.Omega0_m))**(-0.5) #expansion fixed to LCDM
+
+    def beta(self, a):
+        if self.nDGP: return 1. + self.H_NC(a)*(1.+a*self.dHda_NC(a)/3./self.H_NC(a))/np.sqrt(self.Omega_rc)
+        else: return 1.
+
+    def mu(self, a):
+        if self.nDGP: return 1.+1./(3.*self.beta(a))
+        else: return 1.
+
+    def mu2(self, a):
+        if self.nDGP: return -0.5*self.H_NC(a)**2.*(1./(3.*self.beta(a)))**3./self.Omega_rc
+        else: return 0. #think limit here is 0 not 1?
+
+    def mu22(self, a):
+        if self.nDGP: return 0.5*self.H_NC(a)**4.*(1./(3.*self.beta(a)))**5./self.Omega_rc/self.Omega_rc
+        else: return 0. #think limit here is 0 not 1?
+
+    #######################################################
+    ###### Functions to compute MG growth internally ######
+    # How to make these self.func form with odeint(func)? #
+    #######################################################
+
+    def nDGP_H_NC(a, Omega0_m):
+        """Non-Conformal Hubble, H0=1"""
+        #if self.MG: return np.sqrt(Omega0_m/a/a/a + ((1.-Omega0_m)+ 2.*Omega_rc*(np.sqrt((Omega0_m/a/a/a)/Omega_rc + 1.) - 1.)) + Omega_rc) - np.sqrt(Omega_rc) #nDGP+DE
+        #else: return (Omega0_m/a/a/a + (1.-Omega0_m))**0.5 #LCDM
+        return (Omega0_m/a/a/a + (1.-Omega0_m))**0.5 #expansion fixed to LCDM
+
+    def nDGP_dHda_NC(a, Omega0_m):
+        """Derivative of non-Conformal Hubble w.r.t. a, H0=1"""
+        #if self.MG: return (-3.*a**(-1. - 3.*(1 + w))*(1 + w)*(1.-Omega0_m) - 3*Omega0_m*a**(-4))/(2.*np.sqrt((1.-Omega0_m)*a**(-3*(1 + w)) + Omega0_m*a**(-3) + Omega_rc)) #nDGP+DE
+        #else: return 0.5*(-3.*Omega0_m/a/a/a/a)*(Omega0_m/a/a/a + (1.-Omega0_m))**(-0.5) #LCDM
+        return 0.5*(-3.*Omega0_m/a/a/a/a)*(Omega0_m/a/a/a + (1.-Omega0_m))**(-0.5) #expansion fixed to LCDM
+
+    def nDGP_beta(a, Omega0_m, Omega_rc):
+        """beta function for nDGP""" # Omega_rc as described in 1606.02520 for example
+    	return 1. + nDGP_H_NC(a, Omega0_m)*(1.+a*nDGP_dHda_NC(a, Omega0_m)/3./nDGP_H_NC(a, Omega0_m))/np.sqrt(Omega_rc)
+
+    def mu_MG(a, Omega0_m, Omega_rc):
+        #return 1. #GR
+        return 1.+1./(3.*nDGP_beta(a, Omega0_m, Omega_rc)) #nDGP
+        #var1 = (k0/a)**2.
+        #return 1. + var1/(3.*(var1+pow3(omega0/pow3(a)-4.*(omega0-1.))/(2.*p1/pow2(h0)*pow2(4.-3.*omega0)))) #f(R) Hu- Sawicki
+
+    # how to make this use self?
```

Comment:
I think the following is ok. I'm a bit confused the repetitions of the previous functions. Is this because of the exact/LCDM background?

Reply:
Yes, I tried to solve it in Mathematica but it seems that there's no analytic solution. The Hypergeometric solution comes from the wCDM expression of Omega_m (a combination of terms depending on power of the scale factor), while in nDGP there's this \mu(a) function. I have checked the scipy.odeint solution with the Mathematica solution and they agree at the per mille level for the growing mode, but there is still a per cent discrepancy for the decaying mode.

Reply:
I see, that's funny. Do you think it's a numerical issue? If the equations are the same, percent discrepancy can't be there, we should look into that.

Reply:
So the repeated functions are hopefully temporary and can be removed soon depending on the answer to the question below.

My main question is related to calling odeint(compute_primes_MG, ...). [The odeint call](https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.odeint.html) requires that the function compute_primes_MG takes an array y made up of the variables (in our case D and dD/da) as the first argument and then the x variable (in our case scale factor a) as the second argument, and then any extra arguments after that. 

But with the way the other member functions of the GreenFunction class are written is always def func(self, a). So I don't know how to rewrite compute_primes_MG in the correct form to be a member function of the GreenFunction class (i.e. with self as an argument etc). This is what I need help with.

Reply:
I think the way it's written now is ok, or am I missing something?
Any luck with the precision of the numerical integration?

Reply:
I found the correct way to call odeint(compute_primes_MG) while keeping everything a member function of the GreenFunction class and implemented it in 5116324.

Reply:
@guidodamico @MarcoMarinucci 
The problem now is that it complains about integrals being slowly convergent or divergent and as a result the code is very slow for the nDGP case. I'll try to upload a iPython notebook demonstrating this.

Reply:
I added a iPython notebook (check_nDGP.ipynb) in 7c0e024 which shows the numerical solution for the nDGP decay mode is very problematic -- the ratio to LCDM at a=1 is ~-10^9. Note that this is heavily dependent on a_ini used in the numerical solution -- for a_ini=1e-2 instead of a_ini=1e-7 I find the ratio to LCDM at a=1 is ~+20.

Anyway, I think the rest of the Greens function integrals are very slow as a result of these weird nDGP decay values.